### PR TITLE
Upgrade to Asciidoctor.js 2+

### LINKED ISSUE #252: Upgrade to Asciidoctor.js 2.0.0
The final version is not yet available but I've published a second release candidate today. 
A few notes:

* The package was renamed to `asciidoctor`
* The package `asciidoctor` is now a meta-package that contains `@asciidoctor/core` and `@asciidoctor/cli`
* We probably should rename the package to `@asciidoctor/reveal.js-converter`
* We should make sure that the following is working:

```
$ npm i asciidoctor @asciidoctor/reveal.js-converter
$ asciidoctor -r @asciidoctor/reveal.js-converter -b reveal.js presentation.adoc
```

### PR DESCRIPTION
**Pull request by ggrossetie**
Resolves #252

### CODE REVIEW

**Comment by obilodeau** [comment]
I would prefer the unrelated changes to be split out.

File: package.json [hunk 1]
```diff
@@ -1,26 +1,24 @@
...
-    "npm": ">=3.0.0"
+    "node": ">=8.11",
+    "npm": ">=5.0.0"
   },
-  "files": [
-    "dist",
-    "LICENSE.adoc",
-    "README.adoc"
-  ],
+  "files": [],
```
Comment (obilodeau) [inline package.json]: why change this?

File: package.json [hunk 2]
```diff
@@ -41,10 +39,13 @@
   },
   "homepage": "https://github.com/asciidoctor/asciidoctor-reveal.js",
   "dependencies": {
-    "asciidoctor.js": "1.5.9",
-    "reveal.js": "3.7.0"
+    "reveal.js": "3.6.0"
```
Comment (obilodeau) [inline package.json]: Why the version downgrade?

**Review by obilodeau — COMMENTED** [review]
Note: I haven't tested it yet.

It's cool to get rid of the need to have users create their own converter script but to also keep it as an option with a stable API.

**Comment by Mogztter** [comment]
Could you be more specific ? All changes are related to the upgrade 🤔

File: package.json [hunk 3]
```diff
@@ -1,26 +1,24 @@
...
+  "name": "@asciidoctor/reveal.js-converter",
   "version": "2.0.1-dev",
-  "description": "A Reveal.js converter for Asciidoctor and Asciidoctor.js. Write your slides in AsciiDoc!",
+  "description": "A Reveal.js converter for Asciidoctor.js. Write your slides in AsciiDoc!",
   "main": "dist/main.js",
   "engines": {
-    "node": ">=4",
-    "npm": ">=3.0.0"
+    "node": ">=8.11",
+    "npm": ">=5.0.0"
```
[the remark this answers is not in the document]
Comment (Mogztter) [reply package.json]: Asciidoctor.js requires Node.js 8.11 and is not compatible with Node.js 6

File: package.json (hunk 1 above)
Comment (Mogztter) [reply package.json]: By default "LICENSE*", "README*" and the target of `main` are included: https://docs.npmjs.com/files/package.json#files

I can move this change to another pull request, if you want.

File: package.json (hunk 2 above)
Comment (Mogztter) [reply package.json]: Ooops I don't think it's intended... I will revert this change.

**Comment by obilodeau** [comment]
> Could you be more specific ? All changes are related to the upgrade :thinking: 

I would prefer if the new syntax highlighter changes would be separate. I think this is what is breaking the tests right now.

File: package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (obilodeau) [reply package.json]: Surprises me to be specific on `.11`. Ubuntu 18.04 LTS only seem to come with `8.10`.

Anyways, after 5 minutes of googling, there seem to be well-supported snap packages and PPAs for all major distributions with decent install guides: https://github.com/nodesource/distributions/blob/master/README.md.

File: package.json (hunk 1 above)
Comment (obilodeau) [reply package.json]: No, it's good.

**Comment by Mogztter** [comment]
> I would prefer if the new syntax highlighter changes would be separate. I think this is what is breaking the tests right now.

A test was failing when running with Asciidoctor.js: https://github.com/asciidoctor/asciidoctor-reveal.js/blob/master/examples/source-pygments.adoc

But you're right, I'm using the following (which is working with Asciidoctor.js but failing with Asciidoctor):

```slim
- linkcss = node.attr? 'linkcss'
```

I think it should be `- linkcss = attr? 'linkcss'`

**Comment by johnjelinek** [comment]
Is this good to merge?

**Comment by Mogztter** [comment]
@obilodeau Please let me know if there's anything to fix before we can merge this upgrade 😃

File: templates/document.html.slim [hunk 4]
```diff
@@ -46,19 +47,8 @@ html lang=(attr :lang, 'en' unless attr? :nolang)
...
-        - else
-          style=Asciidoctor::Stylesheets.instance.coderay_stylesheet_data
-    - when 'pygments'
-      - if (attr 'pygments-css', 'class') == 'class'
-        - if @safe >= Asciidoctor::SafeMode::SECURE || (attr? :linkcss)
-          link rel='stylesheet' href=normalize_web_path('asciidoctor-pygments.css', (attr :stylesdir, ''))
-        - else
-          style=Asciidoctor::Stylesheets.instance.pygments_stylesheet_data(attr 'pygments-style')
+    - if (syntax_hl = document.syntax_highlighter) && (syntax_hl.docinfo? :head)
+      =(syntax_hl.docinfo :head, node, linkcss: linkcss)
```
Comment (obilodeau) [inline templates/document.html.slim]: Doesn't this remove coderay and pygment support?

File: package.json [hunk 5]
```diff
@@ -41,10 +39,13 @@
   },
   "homepage": "https://github.com/asciidoctor/asciidoctor-reveal.js",
   "dependencies": {
-    "asciidoctor.js": "1.5.9",
     "reveal.js": "3.7.0"
   },
+  "peerDependencies": {
+    "@asciidoctor/core": "^2.0.0"
+  },
```
Comment (obilodeau) [inline package.json]: This move is following this discussion: https://github.com/asciidoctor/asciidoctor-reveal.js/issues/187

File: npm/examples.js [hunk 6]
```diff
@@ -7,7 +7,7 @@ const examplesDir = 'examples';
 log.task('examples');
 
 // Load asciidoctor.js and local asciidoctor-reveal.js
-const asciidoctor = require('asciidoctor.js')();
+const asciidoctor = require('@asciidoctor/core')();
 const asciidoctorRevealjs = require('../build/asciidoctor-reveal.js');
```
Comment (obilodeau) [inline npm/examples.js]: This API change leads me to believe the README also needs updating?

**Review by obilodeau — CHANGES_REQUESTED** [review]

File: package.json (hunk 5 above)
Comment (obilodeau) [reply package.json]: I think this should be documented as a comment. As a casual npm user, the added complexity of peer dependency is unclear. Something should be written to document that choice. I'm thinking specifically of the relationship with Opal versioning here. Some dev docs should need updating too.

**Comment by obilodeau** [comment]
I just did a new review. Only minor stuff, mostly documentation and comments.

File: templates/document.html.slim (hunk 4 above)
Comment (Mogztter) [reply templates/document.html.slim]: No both are working with Asciidoctor Ruby.
The code is using the new syntax highlighter API. The method `document.syntax_highlighter` will get the syntax highlighter configured: https://github.com/asciidoctor/asciidoctor/tree/master/lib/asciidoctor/syntax_highlighter


coderay
pygment

File: npm/examples.js (hunk 6 above)
Comment (Mogztter) [reply npm/examples.js]: Yes :+1:

File: README.adoc [hunk 7]
```diff
@@ -130,38 +130,38 @@ TIP: If you are using https://pages.github.com/[GitHub Pages], plan ahead by kee
...
-// Load asciidoctor.js and asciidoctor-reveal.js
-var asciidoctor = require('asciidoctor.js')();
-var asciidoctorRevealjs = require('asciidoctor-reveal.js');
-asciidoctorRevealjs.register()
+Once the dependencies are installed, verify that the `asciidoctor` command is available by running:
 
-// Convert the document 'presentation.adoc' using the reveal.js converter
-var options = {safe: 'safe', backend: 'revealjs'};
-asciidoctor.convertFile('presentation.adoc', options); // <1>
+ $ $(npm bin)/asciidoctor --version
```
Comment (Mogztter) [inline README.adoc]: I will add a comment for Windows users as `$(npm bin)` won't work.

As an alternative, it's possible to define a task in the package.json:

```json
{
"scripts": {
    "asciidoctor": "asciidoctor"
  }
}
```

```
$ npm run asciidoctor -- --version
```

Or to use a relative path:

```
$ ./node_modules/.bin/asciidoctor --version
```

Please note that `./` is a Unix notation, not sure about the syntax on Windows.

**Comment by Mogztter** [comment]
This pull request has too much changes, I will open pull request with smaller changes so it's easier for you to review.

File: templates/document.html.slim (hunk 4 above)
Comment (Mogztter) [reply templates/document.html.slim]: But it's not compatible with 1.5.8 so it breaks backward compatibility with Asciidoctor < 2.0.0

Comment (Mogztter) [reply templates/document.html.slim]: I guess we could check if the method `document.syntax_highlighter` exists to use the new API or not?

Comment (Mogztter) [reply templates/document.html.slim]: Let's revert this change, we can do a follow-up pull request to support the new syntax highlighter API.

File: package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (Mogztter) [reply package.json]: FYI Node 8 will be EOL on January 2020 :laughing: 
The active LTS is Node 10 or Node 12.

File: README.adoc [hunk 8]
```diff
@@ -130,38 +130,43 @@ TIP: If you are using https://pages.github.com/[GitHub Pages], plan ahead by kee
...
+First you must install and configure {uri-nodejs-download}[Node] on your machine.
 
 === Install
 
-Using npm:
+We recommend to install the dependencies in a project directory, such as the directory where your AsciiDoc presentations are stored.
+If you don't have a `package.json` file in your project directory, you can create one to eliminate warnings during the installation using:
 
-Create a directory to place slides. Initialize the directory to store npm packages within that directory. 
+ $ npm init -y
```
Comment (Mogztter) [inline README.adoc]: Since `echo "{}" > package.json` does not work on Windows I think it's best to use `npm init -y`.

**Comment by Mogztter** [comment]
Now the document is Windows/PowerShell friendly.
I think it's ready to go.

**Review by obilodeau — APPROVED** [review]

**Pull request merged**